File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
from collections import deque
def solution(s):
    st = deque()
    
    for ch in s:
        if ch == "(":
            st.append(ch)
        else:
            if st:
                st.pop()
            else:
                return False
                
    if st:
        return False
    return True

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;

        ArrayDeque<Character> st = new ArrayDeque<>();
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            
            if(c == '('){
                st.addLast(c);
            } 
            else{
                if(st.isEmpty() || st.peekLast() == c){
                    return false;
                }
                else{
                    st.pop();
                }
            }
        }
        
        if(!st.isEmpty()) return false;

        return answer;
    }
}